src/test/regress/sql/distributed_triggers.sql

```diff
@@ -0,0 +1,392 @@
+
+CREATE TRIGGER record_change_trigger
+AFTER INSERT OR UPDATE OR DELETE ON data
+FOR EACH ROW EXECUTE FUNCTION distributed_triggers.record_change();
+
+-- Trigger function should appear on workers
+SELECT proname from pg_proc WHERE oid='distributed_triggers.record_change'::regproc;
+SELECT run_command_on_workers($$SELECT count(*) FROM pg_proc WHERE oid='distributed_triggers.record_change'::regproc$$);
+
+INSERT INTO data VALUES ('hello','world','{"hello":"world"}');
+INSERT INTO data VALUES ('hello2','world2','{"hello2":"world2"}');
+INSERT INTO data VALUES ('hello3','world3','{"hello3":"world3"}');
+DELETE FROM data where shard_key_value = 'hello';
+BEGIN;
+UPDATE data SET value = '{}' where shard_key_value = 'hello3';
+END;
+DELETE FROM data where shard_key_value = 'hello3';
+
+SELECT * FROM data
+ORDER BY shard_key_value, object_id;
+SELECT shard_key_value, object_id, change_id, operation_type, new_value
+FROM data_changes
+ORDER BY shard_key_value, object_id, change_id;
+
+CREATE FUNCTION insert_delete_document(key text, id text)
+RETURNS void LANGUAGE plpgsql AS $fn$
+BEGIN
+	INSERT INTO distributed_triggers.data VALUES (key, id, '{"id1":"id2"}');
+	DELETE FROM distributed_triggers.data where shard_key_value = key;
+END;
+$fn$;
+
+SELECT create_distributed_function(
+  'insert_delete_document(text, text)', 'key',
+  colocate_with := 'data',
+  force_delegation := true
+);
+
+SELECT insert_delete_document('hello4', 'world4');
+BEGIN;
+SELECT insert_delete_document('hello4', 'world4');
+COMMIT;
+
+SELECT * FROM data
+ORDER BY shard_key_value, object_id;
+SELECT shard_key_value, object_id, change_id, operation_type, new_value
+FROM data_changes
+ORDER BY shard_key_value, object_id, change_id;
+DROP TRIGGER record_change_trigger ON data;
+
+--
+-- Run bad triggers
+--
+CREATE OR REPLACE FUNCTION bad_shardkey_record_change()
+RETURNS trigger
+AS $$
+DECLARE
+    last_change_id bigint;
+BEGIN
+    INSERT INTO distributed_triggers.data_changes (shard_key_value, object_id, change_id, operation_type, new_value)
+    VALUES ('BAD', NEW.object_id, COALESCE(last_change_id + 1, 1), TG_OP, NEW.value);
+    RETURN NULL;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER bad_shardkey_record_change_trigger
+AFTER INSERT OR UPDATE OR DELETE ON data
+FOR EACH ROW EXECUTE FUNCTION distributed_triggers.bad_shardkey_record_change();
+
+-- Bad trigger fired from an individual SQL
+-- Query-on-distributed table exception should catch this
+INSERT INTO data VALUES ('hello6','world6','{"hello6":"world6"}');
+
+-- Bad trigger fired from SQL inside a force-delegated function
+-- Incorrect distribution key exception should catch this
+SELECT insert_delete_document('hello6', 'world6');
+
+SELECT * FROM data
+ORDER BY shard_key_value, object_id;
+SELECT shard_key_value, object_id, change_id, operation_type, new_value
+FROM data_changes
+ORDER BY shard_key_value, object_id, change_id;
+DROP TRIGGER bad_shardkey_record_change_trigger ON data;
+
```

Comment:
could be good to add some ALTER TRIGGER test (mainly rename)

manual testing looked fine

Reply:
Done.